v2/tools/generator/internal/codegen/pipeline/repair_skipping_properties_test.go

```diff
@@ -0,0 +1,308 @@
+	)
+	g.Expect(err).To(Succeed())
+
+	// Act - run the Repairer stage
+	finalState, err := RunTestPipeline(
+		initialState,
+		RepairSkippingProperties(), // and then we get to run the stage we're testing
+	)
+	g.Expect(err).To(BeNil())
+
+	// Assert - we expect no error, and no new definitions
+	g.Expect(err).To(BeNil())
+	g.Expect(finalState.definitions).To(HaveLen(10)) // Our five original resources, plus the storage variants
+}
+
+// Test_RepairSkippingProperties checks that the pipeline stage correctly injects new types when properties of different types skip versions.
+func Test_RepairSkippingProperties_WhenPropertyTypesDiffer_InjectsExpectedAdditionalDefinition(t *testing.T) {
+	t.Parallel()
+	g := NewGomegaWithT(t)
+
+	// Arrange - create multiple versions of Person, where the Residence goes missing and returns with a different shape
+	residenceV1 := test.CreateObjectDefinition(test.Pkg2020, "Residence", test.FullAddressProperty, test.CityProperty)
+	residenceV3 := test.CreateObjectDefinition(test.Pkg2022, "Residence", test.FullAddressProperty, test.SuburbProperty, test.CityProperty)
+
+	residenceV1Prop := astmodel.NewPropertyDefinition("Residence", "residence", residenceV1.Name())
+	residenceV3Prop := astmodel.NewPropertyDefinition("Residence", "residence", residenceV3.Name())
+
+	personV1 := test.CreateSpec(test.Pkg2020, "Person", test.FullNameProperty, test.FamilyNameProperty, residenceV1Prop)
+	personV2 := test.CreateSpec(test.Pkg2021, "Person", test.FullNameProperty, test.FamilyNameProperty)
+	personV3 := test.CreateSpec(test.Pkg2022, "Person", test.FullNameProperty, test.FamilyNameProperty, residenceV3Prop)
+
+	defs := make(astmodel.TypeDefinitionSet)
+	defs.AddAll(personV1, personV2, personV3, residenceV1, residenceV3)
+
+	// Arrange - create our initial state, prior to running the Repairer
+	cfg := config.NewConfiguration()
+	initialState, err := RunTestPipeline(
+		NewState(defs),
+		CreateStorageTypes(),            // First create the storage types
+		CreateConversionGraph(cfg, "v"), // Then, create the conversion graph showing relationships
+	)
+	g.Expect(err).To(Succeed())
+
+	// Act - run the Repairer stage
+	finalState, err := RunTestPipeline(
+		initialState,
+		RepairSkippingProperties(), // and then we get to run the stage we're testing
+	)
+
+	// Assert - we expect no error, and one new definition
+	g.Expect(err).To(BeNil())
+	g.Expect(finalState.definitions).To(HaveLen(11)) // Our five original resources, plus the storage variants, plus one more
+
+	expected := astmodel.MakeInternalTypeName(
+		astmodel.MakeCompatPackageReference(test.Pkg2021s),
+		"Residence")
+	g.Expect(finalState.definitions).To(HaveKey(expected))
+}
+
+// Test_RepairSkippingProperties checks that the pipeline stage correctly injects new types with the right functions
+// when properties of different types skip versions.
+func TestGolden_RepairSkippingProperties_WhenPropertyTypesDiffer_GeneratesExpectedCode(t *testing.T) {
```

Comment:
minor: Wondering if it would be clearer if we also had a `repair_skipping_properties_golden_test.go` and put this there, since this file is otherwise pure unit tests?

Don't feel super strongly about this, was just initially confused that you had golden files at all as I didn't see this test when I scanned this file initially and I was wondering where the golden files came from. Only on re-read did I realize that there was one golden test at the end of the file.

Obviously minor.